File: site/07-Graphs-and-Disjoint-Set/java/Dijkstra.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.PriorityQueue;

public class Dijkstra {
    static class Edge {
        int dest, weight;
        Edge(int dest, int weight) {
            this.dest = dest;
            this.weight = weight;
        }
    }

    static class Node {
        int vertex, dist;
        Node(int vertex, int dist) {
            this.vertex = vertex;
            this.dist = dist;
        }
    }

    public static void dijkstra(ArrayList<ArrayList<Edge>> graph, int start, int numVertices) {
        PriorityQueue<Node> pq = new PriorityQueue<>(Comparator.comparingInt(n -> n.dist));
        int[] distances = new int[numVertices];
        Arrays.fill(distances, Integer.MAX_VALUE);

        distances[start] = 0;
        pq.add(new Node(start, 0));

        while (!pq.isEmpty()) {
            Node current = pq.poll();
            int u = current.vertex;

            if (current.dist > distances[u]) continue;

            for (Edge edge : graph.get(u)) {
                int v = edge.dest;
                int weight = edge.weight;

                if (distances[u] + weight < distances[v]) {
                    distances[v] = distances[u] + weight;
                    pq.add(new Node(v, distances[v]));
                }
            }
        }

        System.out.println("Shortest distances from vertex " + start + ":");
        for (int i = 0; i < numVertices; i++) {
            System.out.println("To vertex " + i + ": " + distances[i]);
        }
    }

    public static void main(String[] args) {
        int numVertices = 4;
        ArrayList<ArrayList<Edge>> graph = new ArrayList<>(numVertices);
        for (int i = 0; i < numVertices; i++) {
            graph.add(new ArrayList<>());
        }

        graph.get(0).add(new Edge(1, 4));
        graph.get(0).add(new Edge(2, 1));
        graph.get(1).add(new Edge(3, 1));
        graph.get(2).add(new Edge(1, 2));
        graph.get(2).add(new Edge(3, 5));

        dijkstra(graph, 0, numVertices);
    }
}

File: site/07-Graphs-and-Disjoint-Set/python/dijkstra.py
import heapq

def dijkstra(graph, start, num_vertices):
    # Min-priority queue storing tuples of (distance, vertex)
    pq = []
    distances = {i: float('inf') for i in range(num_vertices)}
    
    distances[start] = 0
    heapq.heappush(pq, (0, start))
    
    while pq:
        current_distance, current_vertex = heapq.heappop(pq)
        
        # If we found a shorter path to current_vertex already, skip it
        if current_distance > distances[current_vertex]:
            continue
            
        for neighbor, weight in graph.get(current_vertex, []):
            distance = current_distance + weight
            
            # If a shorter path is found
            if distance < distances[neighbor]:
                distances[neighbor] = distance
                heapq.heappush(pq, (distance, neighbor))
                
    return distances

if __name__ == "__main__":
    # Graph represented as adjacency list
    # Vertex: list of (neighbor, weight)
    graph = {
        0: [(1, 4), (2, 1)],
        1: [(3, 1)],
        2: [(1, 2), (3, 5)],
        3: []
    }
    
    num_vertices = 4
    start_node = 0
    shortest_paths = dijkstra(graph, start_node, num_vertices)
    
    print(f"Shortest distances from vertex {start_node}:")
    for vertex, dist in shortest_paths.items():
        print(f"To vertex {vertex}: {dist}")
